contracts/zero-ex/contracts/src/features/MultiplexFeature.sol

```diff
@@ -0,0 +1,799 @@
+        if (IERC20TokenV06(inputToken).isTokenETH()) {
+            sandbox.executeSellEthForToken(
+                provider,
+                outputToken,
+                recipient,
+                0,
+                auxiliaryData
+            );
+        } else if (IERC20TokenV06(outputToken).isTokenETH()) {
+            sandbox.executeSellTokenForEth(
+                provider,
+                inputToken,
+                recipient,
+                0,
+                auxiliaryData
+            );
+        } else {
+            sandbox.executeSellTokenForToken(
+                provider,
+                inputToken,
+                outputToken,
+                recipient,
+                0,
+                auxiliaryData
+            );
+        }
+        outputTokenAmount = IERC20TokenV06(outputToken).getTokenBalanceOf(recipient)
+            .safeSub(balanceBefore);
+        emit LiquidityProviderSwap(
+            address(inputToken),
+            address(outputToken),
+            inputTokenAmount,
+            outputTokenAmount,
+            address(provider),
+            recipient
+        );
+        return outputTokenAmount;
+    }
+
+    // Some liquidity sources (e.g. Uniswap, Sushiswap, and PLP) can be passed
+    // a `recipient` parameter so the boguht tokens are transferred to the
+    // `recipient` address rather than `msg.sender`.
+    // Some liquidity sources (also Uniswap, Sushiswap, and PLP incidentally)
+    // support a "transfer then execute" pattern, where the token being sold
+    // can be transferred into the contract before calling a swap function to
+    // execute the trade.
+    // If the current hop in a multi-hop fill satisfies the first condition,
+    // and the next hop satisfies the second condition, the tokens bought
+    // in the current hop can be directly sent to the target contract of
+    // the next hop to save a transfer.
+    function _computeHopRecipient(
+        WrappedMultiHopCall[] memory calls,
+        uint256 i
+    )
+        private
+        view
+        returns (address recipient)
+    {
+        recipient = msg.sender;
+        if (i != calls.length - 1) {
+            WrappedMultiHopCall memory nextCall = calls[i + 1];
+            if (nextCall.selector == this._sellToUniswap.selector) {
+                (address[] memory tokens, bool isSushi) = abi.decode(
+                    nextCall.data,
+                    (address[], bool)
+                );
+                recipient = _computeUniswapPairAddress(tokens[0], tokens[1], isSushi);
+            } else if (nextCall.selector == this._sellToLiquidityProvider.selector) {
+                (recipient, ,) = abi.decode(
+                    nextCall.data,
+                    (address, bytes, uint256)
+                );
+            } else if (nextCall.selector == IEtherTokenV06.withdraw.selector) {
+                recipient = address(this);
```

Comment:
Why? If we just end up transferring to the taker anyway?

Reply:
this is the case where the _next_ call is unwrapping WETH, so we want the current call to transfer the WETH to the EP.

Reply:
oh gotcha